pkg/miniogw/multipart.go

```diff
@@ -0,0 +1,455 @@
+			upload.fail(err)
+		} else {
+			upload.complete(minio.ObjectInfo{
+				Name:        object,
+				Bucket:      bucket,
+				ModTime:     result.Modified,
+				Size:        result.Size,
+				ETag:        result.Checksum,
+				ContentType: result.ContentType,
+				UserDefined: result.UserDefined,
+			})
+		}
+	}()
+
+	return upload.ID, nil
+}
+
+func (s *storjObjects) PutObjectPart(ctx context.Context, bucket, object, uploadID string, partID int, data *hash.Reader) (info minio.PartInfo, err error) {
+	defer mon.Task()(&ctx)(&err)
+
+	uploads := s.storj.multipart
+
+	upload, err := uploads.Get(bucket, object, uploadID)
+	if err != nil {
+		return minio.PartInfo{}, err
+	}
+
+	part, err := upload.Stream.AddPart(partID, data)
+	if err != nil {
+		return minio.PartInfo{}, err
+	}
+
+	err = <-part.Done
+	if err != nil {
+		return minio.PartInfo{}, err
+	}
+
+	partInfo := minio.PartInfo{
+		PartNumber:   part.Number,
+		LastModified: time.Now(),
+		ETag:         data.SHA256HexString(),
+		Size:         atomic.LoadInt64(&part.Size),
+	}
+
+	upload.addCompletedPart(partInfo)
+
+	return partInfo, nil
+}
+
+func (s *storjObjects) AbortMultipartUpload(ctx context.Context, bucket, object, uploadID string) (err error) {
+	defer mon.Task()(&ctx)(&err)
+
+	uploads := s.storj.multipart
+
+	upload, err := uploads.Remove(bucket, object, uploadID)
+	if err != nil {
+		return err
+	}
+
+	errAbort := Error.New("abort")
+	upload.Stream.Abort(errAbort)
+	r := <-upload.Done
+	if r.Error != errAbort {
+		return r.Error
+	}
+	return nil
+}
+
+func (s *storjObjects) CompleteMultipartUpload(ctx context.Context, bucket, object, uploadID string, uploadedParts []minio.CompletePart) (objInfo minio.ObjectInfo, err error) {
+	defer mon.Task()(&ctx)(&err)
+
+	uploads := s.storj.multipart
+	upload, err := uploads.Remove(bucket, object, uploadID)
+	if err != nil {
+		return minio.ObjectInfo{}, err
+	}
+
+	// notify stream that there aren't more parts coming
+	upload.Stream.Close()
+	// wait for completion
+	result := <-upload.Done
+	// return the final info
+	return result.Info, result.Error
+}
+
+func (s *storjObjects) ListObjectParts(ctx context.Context, bucket, object, uploadID string, partNumberMarker int, maxParts int) (result minio.ListPartsInfo, err error) {
+	uploads := s.storj.multipart
+	upload, err := uploads.Get(bucket, object, uploadID)
+	if err != nil {
+		return minio.ListPartsInfo{}, err
+	}
+
+	list := minio.ListPartsInfo{}
+
+	list.Bucket = bucket
+	list.Object = object
+	list.UploadID = uploadID
+	list.PartNumberMarker = partNumberMarker
+	list.MaxParts = maxParts
+	list.UserDefined = upload.Metadata
+	list.Parts = upload.getCompletedParts()
+
+	sort.Slice(list.Parts, func(i, k int) bool {
+		return list.Parts[i].PartNumber < list.Parts[k].PartNumber
+	})
+
+	var first int
+	for i, p := range list.Parts {
+		first = i
+		if partNumberMarker <= p.PartNumber {
+			break
+		}
+	}
+
+	list.Parts = list.Parts[first:]
+	if len(list.Parts) < maxParts {
```

Comment:
Wait... why did you change:
```
if len(list.Parts) > maxParts {
```
to 
```
if len(list.Parts) < maxParts {
```
We should return a truncated result only if the remaining parts in the list are more than `maxParts`. Not the opposite.

Reply:
ah, typo... I tried the other order maxParts < len(...), but that didn't improve readability, so moved it back, but forgot the comparison.